File: python/src/agent_framework/core/provider_summarizer.py
from __future__ import annotations

import json

from agent_framework.core.domain import Message


def _message_to_text(message: Message) -> str:
    if isinstance(message.content, str):
        return message.content
    return json.dumps(message.content)


def create_provider_summarizer(provider):
    async def summarize(messages: list[Message]) -> str:
        transcript = "\n".join(f"{message.role}: {_message_to_text(message)}" for message in messages)
        summary = ""
        async for chunk in provider.stream(
            [
                Message(
                    role="system",
                    content="Summarize the following conversation for future context. Keep key facts and decisions.",
                ),
                Message(role="user", content=transcript),
            ],
            [],
        ):
            if chunk.type == "text_delta" and chunk.text_delta:
                summary += chunk.text_delta
        return summary.strip() or transcript[:500]

    return summarize

File: typescript/packages/core/src/provider-summarizer.ts
import type { ModelProvider } from './types/contracts.js';
import type { Message } from './types/domain.js';
import type { SummarizeMessages } from './llm-summarization-compactor.js';

function messageToText(message: Message): string {
  return typeof message.content === 'string' ? message.content : JSON.stringify(message.content);
}

export function createProviderSummarizer(provider: ModelProvider): SummarizeMessages {
  return async (messages: Message[]) => {
    const transcript = messages.map((message) => `${message.role}: ${messageToText(message)}`).join('\n');
    let summary = '';

    for await (const chunk of provider.stream({
      messages: [
        {
          role: 'system',
          content: 'Summarize the following conversation for future context. Keep key facts and decisions.',
        },
        { role: 'user', content: transcript },
      ],
      tools: [],
    })) {
      if (chunk.type === 'text_delta' && chunk.textDelta) {
        summary += chunk.textDelta;
      }
    }

    return summary.trim() || transcript.slice(0, 500);
  };
}
